core/os/midi_driver.cpp

```diff
@@ -42,77 +42,88 @@ void MIDIDriver::set_singleton() {
+		if (p_length - index >= 1) {
+			if (p_data[index] >= 0xF0) {
+				// Channel does not apply to system common messages.
+				event->set_channel(0);
+				event->set_message(MIDIMessage(p_data[index]));
+				last_received_message = p_data[index];
+				index++;
+			} else if ((p_data[index] & 0x80) == 0x00) {
+				// Running status, index is not changed because no new MIDI status byte, just additional data bytes.
+				event->set_channel(last_received_message & 0xF);
+				event->set_message(MIDIMessage(last_received_message >> 4));
+			} else {
+				event->set_channel(p_data[index] & 0xF);
+				event->set_message(MIDIMessage(p_data[index] >> 4));
+				last_received_message = p_data[index];
+				index++;
 			}
-			break;
+		}
 
-		case MIDIMessage::NOTE_ON:
-		case MIDIMessage::NOTE_OFF:
-			if (length >= 2 + param_position) {
-				event->set_pitch(data[param_position]);
-				event->set_velocity(data[param_position + 1]);
-			}
-			break;
+		switch (event->get_message()) {
+			case MIDIMessage::AFTERTOUCH:
+				if (p_length >= 2 + index) {
+					event->set_pitch(p_data[index++]);
+					event->set_pressure(p_data[index++]);
+				}
+				break;
 
-		case MIDIMessage::PITCH_BEND:
-			if (length >= 2 + param_position) {
-				event->set_pitch((data[param_position + 1] << 7) | data[param_position]);
-			}
-			break;
+			case MIDIMessage::CONTROL_CHANGE:
+				if (p_length >= 2 + index) {
+					event->set_controller_number(p_data[index++]);
+					event->set_controller_value(p_data[index++]);
+				}
+				break;
 
-		case MIDIMessage::PROGRAM_CHANGE:
-			if (length >= 1 + param_position) {
-				event->set_instrument(data[param_position]);
-			}
-			break;
+			case MIDIMessage::NOTE_ON:
+			case MIDIMessage::NOTE_OFF:
+				if (p_length >= 2 + index) {
+					event->set_pitch(p_data[index++]);
+					event->set_velocity(p_data[index++]);
+				}
+				break;
 
-		case MIDIMessage::CHANNEL_PRESSURE:
-			if (length >= 1 + param_position) {
-				event->set_pressure(data[param_position]);
-			}
-			break;
-		default:
-			break;
-	}
+			case MIDIMessage::PITCH_BEND:
+				if (p_length >= 2 + index) {
+					event->set_pitch((p_data[index] << 7) | p_data[index]);
+					index++;
+				}
+				break;
+
+			case MIDIMessage::PROGRAM_CHANGE:
+				if (p_length >= 1 + index) {
+					event->set_instrument(p_data[index++]);
+				}
+				break;
+
+			case MIDIMessage::CHANNEL_PRESSURE:
+				if (p_length >= 1 + index) {
+					event->set_pressure(p_data[index++]);
+				}
+				break;
 
-	Input *id = Input::get_singleton();
-	id->parse_input_event(event);
+			default:
+				// Default case means bytes remain that are not yet turned into an InputEventMIDI,
+				// either because its MIDI that this code does not support or a malformed MIDI packet
+				// from an upstream device. This would cause an infinite loop of the while.
+				// To prevent hanging, remaining bytes in p_data are ignored.
+				// (Which is what this method did in previous versions).
```

Comment:
This message is a bit too verbose, I'd just say something like "Anything else means an invalid packet, exit.", we might want to add an error message here, unsure

Reply:
There is `WARN_PRINT_ONCE`, but it depends whether it can get hit in non-error conditions (e.g. unrecognised event).